src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/KeyboardEventSender.cs

```diff
@@ -24,83 +25,101 @@ public static void Reset()
-                        // convert key to text
-                        if (KeyboardInputActionObserver.KeyboardKeyToValueMap.TryGetValue(((KeyControl)inputControl).keyCode, out var possibleValues))
+                        inputControl.WriteValueIntoEvent(upOrDown == KeyState.Down ? 1f : 0f, deltaStateEvent.Value);
+
+                        if (upOrDown == KeyState.Down)
                         {
-                            var value = _isShiftDown ? possibleValues.Item2 : possibleValues.Item1;
-                            if (value == 0x00)
+                            // send a text event so that 'onChange' text events fire
+                            // convert key to text
+                            if (KeyboardInputActionObserver.KeyboardKeyToValueMap.TryGetValue(((KeyControl)inputControl).keyCode, out var possibleValues))
                             {
-                                RGDebug.LogError($"Found null value for keyboard input {key}");
-                                return;
+                                var value = _isShiftDown ? possibleValues.Item2 : possibleValues.Item1;
+                                if (value == 0x00)
+                                {
+                                    RGDebug.LogError($"Found null value for keyboard input {key}");
+                                    return;
+                                }
+
+                                var inputEvent = TextEvent.Create(Keyboard.current.deviceId, value, time);
+                                RGDebug.LogInfo($"({replaySegment}) Sending Text Event for char: '{value}'");
+                                InputSystem.QueueEvent(ref inputEvent);
                             }
-
-                            var inputEvent = TextEvent.Create(Keyboard.current.deviceId, value, time);
-                            RGDebug.LogInfo($"({replaySegment}) Sending Text Event for char: '{value}'");
-                            InputSystem.QueueEvent(ref inputEvent);
                         }
                     }
+
                 }
 
+                if (deltaStateEvent.HasValue)
+                {
+                    RGDebug.LogInfo($"({replaySegment}) Sending Multiple Key Event: [" + string.Join(", ", logList.Select(a => a.Item1 + ":" + a.Item2).ToArray()) + "]");
+                    InputSystem.QueueEvent(deltaStateEvent.Value);
+                    logList.Clear();
+                }
             }
-
-            if (deltaStateEvent.HasValue)
+            finally
             {
-                RGDebug.LogInfo($"({replaySegment}) Sending Multiple Key Event: [" + string.Join(", ", logList.Select(a => a.Item1 + ":" + a.Item2).ToArray()) + "]");
-                InputSystem.QueueEvent(deltaStateEvent.Value);
-                logList.Clear();
+                if (deltaStateEventArray.HasValue)
+                {
+                    deltaStateEventArray.Value.Dispose();
```

Comment:
could this possibly dispose the event before the input system processes the queued event ?